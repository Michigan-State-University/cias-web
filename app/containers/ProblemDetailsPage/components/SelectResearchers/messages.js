/*
 * SelectResearchersModal Messages
 *
 * This contains all the text for the SelectResearchersModal component.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.components.SelectResearchersModal';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the SelectResearchersModal component!',
  },
  find: {
    id: `${scope}.find`,
    defaultMessage: 'Find researcher',
  },
  send: {
    id: `${scope}.send`,
    defaultMessage: 'Send copies',
  },
  name: {
    id: `${scope}.name`,
    defaultMessage: 'Name',
  },
  email: {
    id: `${scope}.email`,
    defaultMessage: 'Email',
  },
});
