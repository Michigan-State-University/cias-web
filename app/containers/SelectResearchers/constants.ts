export const TABLE_MAX_HEIGHT = '400px';
