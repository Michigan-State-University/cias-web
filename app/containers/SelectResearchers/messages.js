import { defineMessages } from 'react-intl';

export const scope = 'app.components.SelectResearchers';

export default defineMessages({
  name: {
    id: `${scope}.name`,
    defaultMessage: 'Name',
  },
  email: {
    id: `${scope}.email`,
    defaultMessage: 'Email',
  },
  waitingForActivation: {
    id: `${scope}.waitingForActivation`,
    defaultMessage: 'Waiting for activation',
  },
  success: {
    id: `${scope}.success`,
    defaultMessage: 'Success',
  },
  byEmail: {
    id: `${scope}.byEmail`,
    defaultMessage: 'by email',
  },
  yourTeamMembers: {
    id: `${scope}.yourTeamMembers`,
    defaultMessage: 'Your team members',
  },
});
