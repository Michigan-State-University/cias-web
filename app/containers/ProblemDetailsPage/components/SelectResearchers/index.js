/**
 *
 * SelectResearchers
 *
 */

import React, { useState, useLayoutEffect } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import remove from 'lodash/remove';

import useFilter from 'utils/useFilter';
import Row from 'components/Row';
import Loader from 'components/Loader';
import Column from 'components/Column';
import Box from 'components/Box';
import Text from 'components/Text';
import { StripedTR, Table, TBody, TD, TH, THead } from 'components/Table';
import Checkbox from 'components/Checkbox';
import { StyledButton } from 'components/Button/StyledButton';
import ErrorAlert from 'components/ErrorAlert';
import SearchInput from 'components/Input/SearchInput';
import { useInjectReducer } from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';
import { Roles } from 'models/User/UserRoles';
import {
  fetchUsers,
  makeSelectUserList,
  UserListReducer,
  userListSaga,
} from 'global/reducers/userList';
import { copyProblemRequest } from 'global/reducers/problems';

import { makeSelectUser } from 'global/reducers/auth';
import messages from './messages';

const SelectResearchers = ({
  intl: { formatMessage },
  userList: { users, usersLoading, usersError },
  fetchUsersRequest,
  copyProblemToResearchers,
  problemId,
  onClose,
  user: { id: currentUserId },
}) => {
  useInjectReducer({ key: 'userList', reducer: UserListReducer });

  const [selected, setSelected] = useState([]);

  const [finalUsers, filterValue, setFilterValue] = useFilter(
    users.filter(({ id }) => id !== currentUserId),
    'full_name',
  );
  useLayoutEffect(() => {
    fetchUsersRequest([Roles.researcher]);
  }, []);

  const handleSend = () => {
    copyProblemToResearchers({ problemId, users: selected });
    onClose();
  };

  if (usersLoading)
    return (
      <Box>
        <Loader type="inline" />
      </Box>
    );
  if (usersError)
    return (
      <Box>
        <ErrorAlert errorText={usersError} />
      </Box>
    );
  return (
    <Box>
      <Row pt={10} width="100%">
        <SearchInput
          ml={5}
          width="100%"
          placeholder={formatMessage(messages.find)}
          value={filterValue}
          onChange={e => setFilterValue(e.target.value)}
        />
        <StyledButton width={200} ml={10} onClick={handleSend}>
          <FormattedMessage {...messages.send} />
        </StyledButton>
      </Row>
      <Table>
        <THead>
          <StripedTR>
            <TH scope="col">
              <Column width={300} pl={10}>
                <Text textAlign="left" fontWeight="bold">
                  <FormattedMessage {...messages.name} />
                </Text>
              </Column>
            </TH>
            <TH scope="col">
              <Column width={300}>
                <Text textAlign="left" fontWeight="bold">
                  <FormattedMessage {...messages.email} />
                </Text>
              </Column>
            </TH>
            <TH scope="col">
              <Column width="100%" />
            </TH>
          </StripedTR>
        </THead>
        <TBody>
          {finalUsers &&
            finalUsers.map(user => {
              const { fullName, email, id } = user;
              const isChecked = selected.includes(id);
              const handleClick = () => {
                if (!isChecked) setSelected([...selected, id]);
                else setSelected(remove(selected, elem => elem !== id));
              };
              return (
                <StripedTR key={`row-th-${id}`}>
                  <TD pl={10}>{fullName}</TD>
                  <TD>{email}</TD>
                  <TD pr={10}>
                    <Checkbox checked={isChecked} onClick={handleClick} />
                  </TD>
                </StripedTR>
              );
            })}
        </TBody>
      </Table>
    </Box>
  );
};

SelectResearchers.propTypes = {
  intl: PropTypes.object,
  userList: PropTypes.object,
  user: PropTypes.object,
  fetchUsersRequest: PropTypes.func.isRequired,
  copyProblemToResearchers: PropTypes.func,
  problemId: PropTypes.string,
  onClose: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  userList: makeSelectUserList(),
  user: makeSelectUser(),
});

const mapDispatchToProps = {
  fetchUsersRequest: fetchUsers,
  copyProblemToResearchers: copyProblemRequest,
};

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

const withSaga = injectSaga({ key: 'userList', saga: userListSaga });

export default compose(
  injectIntl,
  withConnect,
  withSaga,
)(SelectResearchers);
