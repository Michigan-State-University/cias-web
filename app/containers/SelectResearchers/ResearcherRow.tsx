import React from 'react';
import { useIntl } from 'react-intl';

import { colors, themeColors } from 'theme';

import InviteIcon from 'assets/svg/invite.svg';
import CheckGreenIcon from 'assets/svg/check-green.svg';

import { UserItemState, UserWithState } from 'global/reducers/userList';

import { StripedTR, NoMaxWidthTD } from 'components/Table';
import Row from 'components/Row';
import { EllipsisText } from 'components/Text';
import { ImageButton } from 'components/Button';

import messages from './messages';

type Props = {
  researcher: UserWithState;
  onSelected: () => void;
  actionName: string;
};

const ResearcherRow = ({
  researcher: { fullName, email, state },
  onSelected,
  actionName,
}: Props) => {
  const { formatMessage } = useIntl();

  const trimmedFullName = fullName.trim();

  return (
    <StripedTR
      height={53}
      stripesPlacement="odd"
      color={colors.aliceBlueSaturated}
      bg={colors.white}
    >
      <NoMaxWidthTD padding={8}>
        <EllipsisText
          text={trimmedFullName || formatMessage(messages.waitingForActivation)}
          color={!trimmedFullName && themeColors.warning}
        />
      </NoMaxWidthTD>
      <NoMaxWidthTD padding={8}>
        <EllipsisText text={email} />
      </NoMaxWidthTD>
      <NoMaxWidthTD pl={8}>
        <Row>
          {(!state ||
            state === UserItemState.IDLE ||
            state === UserItemState.LOADING) && (
            <ImageButton
              src={InviteIcon}
              onClick={onSelected}
              title={actionName}
              loading={state === UserItemState.LOADING}
              spinnerProps={{ margin: 'initial', color: colors.black }}
            />
          )}
          {state === UserItemState.SUCCESS && (
            <ImageButton
              src={CheckGreenIcon}
              title={formatMessage(messages.success)}
            />
          )}
        </Row>
      </NoMaxWidthTD>
    </StripedTR>
  );
};

export default ResearcherRow;
