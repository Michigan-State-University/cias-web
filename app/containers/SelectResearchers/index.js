/**
 *
 * SelectResearchers
 *
 */

import React, { useState, useLayoutEffect } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import remove from 'lodash/remove';
import trim from 'lodash/trim';

import { colors } from 'theme';
import useFilter from 'utils/useFilter';
import Row from 'components/Row';
import Loader from 'components/Loader';
import Column from 'components/Column';
import Box from 'components/Box';
import Text from 'components/Text';
import { StripedTR, Table, TBody, TD, TH, THead } from 'components/Table';
import Checkbox from 'components/Checkbox';
import { StyledButton } from 'components/Button/StyledButton';
import ErrorAlert from 'components/ErrorAlert';
import SearchInput from 'components/Input/SearchInput';
import { injectSaga, injectReducer } from 'redux-injectors';

import { Roles } from 'models/User/UserRoles';
import {
  makeSelectUserList,
  UserListReducer,
  userListSaga,
  fetchResearchersRequest,
} from 'global/reducers/userList';

import { makeSelectUser } from 'global/reducers/auth';
import { ternary } from 'utils/ternary';
import messages from './messages';

const SelectResearchers = ({
  intl: { formatMessage },
  userList: {
    researchersSelector,
    researchersSelectorLoading,
    researchersSelectorError,
  },
  fetchUsersRequest,
  onClose,
  user: { id: currentUserId },
  onResearchersSelected,
}) => {
  const [selected, setSelected] = useState([]);

  const [finalUsers, filterValue, setFilterValue] = useFilter(
    researchersSelector.filter(({ id }) => id !== currentUserId),
    'fullName',
    {},
  );
  useLayoutEffect(() => {
    fetchUsersRequest([Roles.researcher]);
  }, []);

  const handleSend = () => {
    onResearchersSelected(selected);
    onClose();
  };

  const getDisplayName = fullName => {
    const trimmedFullName = trim(fullName);
    return ternary(
      trimmedFullName,
      trimmedFullName,
      <Text color={colors.flamingo}>
        {formatMessage(messages.waitingForActivation)}
      </Text>,
    );
  };

  if (researchersSelectorLoading)
    return (
      <Box>
        <Loader type="inline" />
      </Box>
    );
  if (researchersSelectorError)
    return (
      <Box>
        <ErrorAlert errorText={researchersSelectorError} />
      </Box>
    );
  return (
    <Box>
      <Row pt={10} width="100%">
        <SearchInput
          ml={5}
          width="100%"
          placeholder={formatMessage(messages.find)}
          value={filterValue}
          onChange={e => setFilterValue(e.target.value)}
        />
        <StyledButton width={200} ml={10} onClick={handleSend}>
          <FormattedMessage {...messages.send} />
        </StyledButton>
      </Row>
      <Table>
        <THead>
          <StripedTR>
            <TH scope="col">
              <Column width={300} pl={10}>
                <Text textAlign="left" fontWeight="bold">
                  <FormattedMessage {...messages.name} />
                </Text>
              </Column>
            </TH>
            <TH scope="col">
              <Column width={300}>
                <Text textAlign="left" fontWeight="bold">
                  <FormattedMessage {...messages.email} />
                </Text>
              </Column>
            </TH>
            <TH scope="col">
              <Column width="100%" />
            </TH>
          </StripedTR>
        </THead>
        <TBody>
          {finalUsers &&
            finalUsers.map(user => {
              const { fullName, email, id } = user;
              const isChecked = selected.includes(id);
              const handleClick = () => {
                if (!isChecked) setSelected([...selected, id]);
                else setSelected(remove(selected, elem => elem !== id));
              };
              return (
                <StripedTR key={`row-th-${id}`}>
                  <TD pl={10}>{getDisplayName(fullName)}</TD>
                  <TD>{email}</TD>
                  <TD pr={10}>
                    <Checkbox checked={isChecked} onClick={handleClick} />
                  </TD>
                </StripedTR>
              );
            })}
        </TBody>
      </Table>
    </Box>
  );
};

SelectResearchers.propTypes = {
  intl: PropTypes.object,
  userList: PropTypes.object,
  user: PropTypes.object,
  fetchUsersRequest: PropTypes.func.isRequired,
  copyInterventionToResearchers: PropTypes.func,
  interventionId: PropTypes.string,
  onClose: PropTypes.func,
  onResearchersSelected: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  userList: makeSelectUserList(),
  user: makeSelectUser(),
});

const mapDispatchToProps = {
  fetchUsersRequest: fetchResearchersRequest,
};

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

const withReducer = injectReducer({
  key: 'userList',
  reducer: UserListReducer,
});
const withSaga = injectSaga({ key: 'userList', saga: userListSaga });

export default compose(
  injectIntl,
  withConnect,
  withReducer,
  withSaga,
)(SelectResearchers);
