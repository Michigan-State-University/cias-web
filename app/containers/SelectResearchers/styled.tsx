// @ts-ignore
import styled from 'styled-components';

import { TD } from 'components/Table';

// @ts-ignore
export const StyledTD = styled(TD)`
  max-width: 0;
`;
