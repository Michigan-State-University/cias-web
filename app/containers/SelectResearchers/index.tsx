/**
 *
 * SelectResearchers
 *
 */

import React, {
  useState,
  useLayoutEffect,
  FC,
  useMemo,
  ChangeEvent,
} from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';
import { useInjectReducer, useInjectSaga } from 'redux-injectors';

import Search from 'assets/svg/search.svg';

import { themeColors } from 'theme';

import { ApiError } from 'models/Api';
import { User } from 'models/User';

import { emailValidator } from 'utils/validators';

import {
  makeSelectUserList,
  UserListReducer,
  userListSaga,
  fetchResearchersRequest,
} from 'global/reducers/userList';

import Row from 'components/Row';
import Box from 'components/Box';
import Text from 'components/Text';
import { Table, TBody, TH, THead, TR } from 'components/Table';
import ErrorAlert from 'components/ErrorAlert';
import Icon from 'components/Icon';
import Button from 'components/Button';
import {
  AdornmentType,
  InputWithAdornment,
} from 'components/Input/InputWithAdornment';
import Spinner from 'components/Spinner';

import messages from './messages';
import { TABLE_MAX_HEIGHT } from './constants';
import ResearcherRow from './ResearcherRow';

export type Props = {
  excludedUserIds?: string[];
  onResearchersSelected: (emails: string[], ids?: string[]) => void;
  actionName: string;
};

const SelectResearchers: FC<Props> = ({
  onResearchersSelected,
  excludedUserIds = [],
  actionName,
}) => {
  const dispatch = useDispatch();
  const { formatMessage } = useIntl();

  useInjectReducer({
    key: 'userList',
    // @ts-ignore
    reducer: UserListReducer,
  });
  useInjectSaga({ key: 'userList', saga: userListSaga });

  const {
    researchersSelectorLoading,
    researchersSelector,
    researchersSelectorError,
  }: {
    researchersSelectorLoading: boolean;
    researchersSelector: (User & { loading: boolean })[];
    researchersSelectorError: ApiError;
  } = useSelector(makeSelectUserList());

  const researchersToShow = useMemo(
    () => researchersSelector.filter(({ id }) => !excludedUserIds.includes(id)),
    [researchersSelector, excludedUserIds],
  );

  useLayoutEffect(() => {
    dispatch(fetchResearchersRequest());
  }, []);

  const [email, setEmail] = useState('');
  const [sendEnabled, setSendEnabled] = useState(false);

  const onInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setEmail(e.target.value);
    setSendEnabled(emailValidator(e.target.value));
  };

  const submit = () => {
    onResearchersSelected([email]);
    setEmail('');
    setSendEnabled(false);
  };

  if (researchersSelectorError)
    return (
      <Box>
        <ErrorAlert errorText={researchersSelectorError} />
      </Box>
    );
  return (
    <Box>
      <Row>
        <InputWithAdornment
          value={email}
          adornmentType={AdornmentType.PREFIX}
          adornment={<Icon src={Search} />}
          onChange={onInputChange}
          placeholder={`${actionName} ${formatMessage(messages.byEmail)}`}
          mr={16}
          adornmentWidth={36}
        />
        <Button disabled={!sendEnabled} width={200} onClick={submit}>
          {actionName}
        </Button>
      </Row>
      {researchersSelectorLoading && (
        <Box mt={16}>
          <Spinner color={themeColors.secondary} />
        </Box>
      )}
      {researchersSelectorError && (
        <Box mt={16}>
          <ErrorAlert errorText={researchersSelectorError} />
        </Box>
      )}
      {!researchersSelectorLoading && !researchersSelectorError && (
        <Box overflow="auto" maxHeight={TABLE_MAX_HEIGHT}>
          <Table width="100%">
            <THead>
              <TR height={46}>
                <TH padding={8}>
                  <Text textAlign="left" fontWeight="bold">
                    <FormattedMessage {...messages.name} />
                  </Text>
                </TH>
                <TH padding={8}>
                  <Text textAlign="left" fontWeight="bold">
                    <FormattedMessage {...messages.email} />
                  </Text>
                </TH>
                <TH padding={8} width={40}>
                  <Text textAlign="left" fontWeight="bold">
                    {actionName}
                  </Text>
                </TH>
              </TR>
            </THead>
            <TBody>
              {researchersToShow.map((researcher) => (
                <ResearcherRow
                  key={researcher.id}
                  researcher={researcher}
                  onSelected={() =>
                    onResearchersSelected([researcher.email], [researcher.id])
                  }
                  actionName={actionName}
                />
              ))}
            </TBody>
          </Table>
        </Box>
      )}
    </Box>
  );
};

export default SelectResearchers;
