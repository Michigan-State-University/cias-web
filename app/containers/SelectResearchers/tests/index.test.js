/**
 *
 * Tests for SelectResearchersModal
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import 'jest-styled-components';
import { DEFAULT_LOCALE } from 'i18n';
import { Provider } from 'react-redux';

import { UserListReducer } from 'global/reducers/userList';
import { authReducer } from 'global/reducers/auth';
import { createTestStore } from 'utils/testUtils/storeUtils';
import SelectResearchers from '../index';

describe('<SelectResearchers />', () => {
  let store;
  const initialState = {
    auth: {
      user: { id: 1 },
    },
  };
  beforeAll(() => {
    store = createTestStore(initialState);
    store.runSaga = () => {};
    store.injectedReducers = {
      problem: UserListReducer,
      auth: authReducer,
    };
    store.injectedSagas = {};
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <SelectResearchers />
        </IntlProvider>
      </Provider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild: renderedComponent },
    } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <Provider store={store}>
          <SelectResearchers />
        </Provider>
      </IntlProvider>,
    );

    expect(renderedComponent).toMatchSnapshot();
  });
});
