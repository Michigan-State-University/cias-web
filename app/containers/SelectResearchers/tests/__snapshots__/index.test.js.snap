// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectResearchers /> Should render and match the snapshot 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: auto;
  width: 50px;
  height: 50px;
}

.c2:after {
  content: ' ';
  display: block;
  width: 33.33333333333333px;
  height: 33.33333333333333px;
  border-radius: 50%;
  border: 6px solid #B727EA;
  border-color: #B727EA transparent #B727EA transparent;
  -webkit-animation: hAjVGa 1.2s linear infinite;
  animation: hAjVGa 1.2s linear infinite;
}

.c1 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  color: unset;
}

<div
  class="c0"
>
  <div
    class="c1"
    data-testid="inline-loader"
  >
    <div
      class="c2"
      color="#B727EA"
      size="50"
      width="6"
    />
  </div>
</div>
`;
